From patchwork Wed Mar 31 03:19:34 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Sean Christopherson <seanjc@google.com>
X-Patchwork-Id: 12174261
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-26.3 required=3.0 tests=BAYES_00,DKIMWL_WL_MED,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	USER_AGENT_GIT,USER_IN_DEF_DKIM_WL autolearn=unavailable autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 3903BC433E1
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 03:20:20 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 14106619BD
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 03:20:20 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S233482AbhCaDTt (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 30 Mar 2021 23:19:49 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:42336 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S233368AbhCaDTp (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 30 Mar 2021 23:19:45 -0400
Received: from mail-yb1-xb4a.google.com (mail-yb1-xb4a.google.com
 [IPv6:2607:f8b0:4864:20::b4a])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 7BC15C061574
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 20:19:45 -0700 (PDT)
Received: by mail-yb1-xb4a.google.com with SMTP id k189so773429ybb.17
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 20:19:45 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=reply-to:date:in-reply-to:message-id:mime-version:references
         :subject:from:to:cc;
        bh=VqSy53IyDnNL/CdQDhT4nMQIBeIb4g0Xh5NCqPSTxiY=;
        b=L3vxRku6tHP9GMz0/JT4UHLAicdp9dcZ+g2J1/dysm0wZpJgxaAjfxTZRGvOqRDoJ5
         5GG5sUkB9HTsjhb7ZcW+2cqWXHS02dQP1cU9Jq70424h6LdhCvBWTWi2Qkb61lFbCmHO
         QF6c6Ih9rAr2eQdCd2pJtRZYC0dBaI0kSCdr+sMMgAg5/lzXj6+0d07tR1OA1a34arnU
         0DF79I759sgJdeCnkcU7kKJCHSg3fuFJRaVtnpmrXCuZnuaRW3w4dGNmlSBalHYxoaW9
         BhI2lgbinJfb1IVflxdD8n2URlBSoiGN/Kq7x1JDvwfuCVb0nwKe97+7cLSKniU4BIdS
         E+2g==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:reply-to:date:in-reply-to:message-id
         :mime-version:references:subject:from:to:cc;
        bh=VqSy53IyDnNL/CdQDhT4nMQIBeIb4g0Xh5NCqPSTxiY=;
        b=SP32Lk8AKf2UMaqNSJZV4HMbzup++f89Fhu8LsEkjQBPfUe+bvo/daxbTJ50xglTzI
         8zOuh++kPoleuNFfN5Pd5ueuK6sBz+TDQT4WqiISHhYVBLdWY5rLV3m2SA6BHHrGd3fr
         6HfkhDqAG2KqjDu+/ClZZvCZygQlscaz3sdcSzxOYlFlG5tN1tQxHuJcfy/XuOzkwLgx
         cSNrXzdBsgRkwinUZMUea9FQAm92mmTqmKiMEnt0GmJPb5HjKw2DUgyfPske0ywgQ5n6
         PgP3LP09iulKLwI750TyrvFqpUzSqxfA8f2bs9DRTlM96g9shw00uvWKhX4+c2Re+z2B
         MdQg==
X-Gm-Message-State: AOAM531Rex02CDk/DGgevJruoksRF6pVOis3q8PXfmERhXzArMwdpmDV
        TUJDryvvjy0NIS/HNIAk/gRn6s327V8=
X-Google-Smtp-Source: 
 ABdhPJwPfX1ZIe9KPSZ+B/9YfLJdMnVqXzEt0iRHW7WxV7gQCnqgXa6cZFbf1Pufpv66ADvtJZu7i2zqXv8=
X-Received: from seanjc798194.pdx.corp.google.com
 ([2620:15c:f:10:6c6b:5d63:9b3b:4a77])
 (user=seanjc job=sendgmr) by 2002:a25:9ac6:: with SMTP id
 t6mr1707731ybo.287.1617160784800;
 Tue, 30 Mar 2021 20:19:44 -0700 (PDT)
Reply-To: Sean Christopherson <seanjc@google.com>
Date: Tue, 30 Mar 2021 20:19:34 -0700
In-Reply-To: <20210331031936.2495277-1-seanjc@google.com>
Message-Id: <20210331031936.2495277-2-seanjc@google.com>
Mime-Version: 1.0
References: <20210331031936.2495277-1-seanjc@google.com>
X-Mailer: git-send-email 2.31.0.291.g576ba9dcdaf-goog
Subject: [PATCH 1/3] KVM: SVM: Use online_vcpus, not created_vcpus, to iterate
 over vCPUs
From: Sean Christopherson <seanjc@google.com>
To: Paolo Bonzini <pbonzini@redhat.com>
Cc: Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org,
        Brijesh Singh <brijesh.singh@amd.com>,
        Tom Lendacky <thomas.lendacky@amd.com>
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Use the kvm_for_each_vcpu() helper to iterate over vCPUs when encrypting
VMSAs for SEV, which effectively switches to use online_vcpus instead of
created_vcpus.  This fixes a possible null-pointer dereference as
created_vcpus does not guarantee a vCPU exists, since it is updated at
the very beginning of KVM_CREATE_VCPU.  created_vcpus exists to allow the
bulk of vCPU creation to run in parallel, while still correctly
restricting the max number of max vCPUs.

Fixes: ad73109ae7ec ("KVM: SVM: Provide support to launch and run an SEV-ES guest")
Cc: stable@vger.kernel.org
Cc: Brijesh Singh <brijesh.singh@amd.com>
Cc: Tom Lendacky <thomas.lendacky@amd.com>
Signed-off-by: Sean Christopherson <seanjc@google.com>
---
 arch/x86/kvm/svm/sev.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/arch/x86/kvm/svm/sev.c b/arch/x86/kvm/svm/sev.c
index 83e00e524513..6481d7165701 100644
--- a/arch/x86/kvm/svm/sev.c
+++ b/arch/x86/kvm/svm/sev.c
@@ -564,6 +564,7 @@ static int sev_launch_update_vmsa(struct kvm *kvm, struct kvm_sev_cmd *argp)
 {
 	struct kvm_sev_info *sev = &to_kvm_svm(kvm)->sev_info;
 	struct sev_data_launch_update_vmsa *vmsa;
+	struct kvm_vcpu *vcpu;
 	int i, ret;
 
 	if (!sev_es_guest(kvm))
@@ -573,8 +574,8 @@ static int sev_launch_update_vmsa(struct kvm *kvm, struct kvm_sev_cmd *argp)
 	if (!vmsa)
 		return -ENOMEM;
 
-	for (i = 0; i < kvm->created_vcpus; i++) {
-		struct vcpu_svm *svm = to_svm(kvm->vcpus[i]);
+	kvm_for_each_vcpu(i, vcpu, kvm) {
+		struct vcpu_svm *svm = to_svm(vcpu);
 
 		/* Perform some pre-encryption checks against the VMSA */
 		ret = sev_es_sync_vmsa(svm);

From patchwork Wed Mar 31 03:19:35 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Sean Christopherson <seanjc@google.com>
X-Patchwork-Id: 12174263
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-26.3 required=3.0 tests=BAYES_00,DKIMWL_WL_MED,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	USER_AGENT_GIT,USER_IN_DEF_DKIM_WL autolearn=ham autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 3D875C433C1
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 03:20:53 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 1B9B9619D8
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 03:20:53 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S233595AbhCaDUU (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 30 Mar 2021 23:20:20 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:42350 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S233448AbhCaDTr (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 30 Mar 2021 23:19:47 -0400
Received: from mail-yb1-xb49.google.com (mail-yb1-xb49.google.com
 [IPv6:2607:f8b0:4864:20::b49])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id A9704C06175F
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 20:19:47 -0700 (PDT)
Received: by mail-yb1-xb49.google.com with SMTP id o129so745102ybg.23
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 20:19:47 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=reply-to:date:in-reply-to:message-id:mime-version:references
         :subject:from:to:cc;
        bh=AxjHjFrYfKI0arB2F1ZWltZrEdR0g/4Qm9dbdrZzkxk=;
        b=LKQVUEz7W1hc1V11sLG/5o9F1h5djL8UVdiWgm2BJ4XSNQXtcBrgZ5NVoEUNtKU+wl
         V+mf1ZEX5NA4i9zJnPvje9Kkx0yg/Ag1ppVoKFk+B5TR+/hlh6SqPvjhjyTmY/5rZTe/
         Y1Cx+BrzkWXpQ6kyjrbc/fgjjmR4TAJPmv37ZXE36/byW9TUXDrjHdqcV0aIEJzYPnMJ
         xp9TtdADSwQLgeQp/EQHlji+TnzWlZ3IEcwlU73/crBG59StMp1e5lAXgNzVYO25R0hY
         mDIDisw5ERtFfPVsT9kZO06WR6qNv6ZOloumT7jW4vbNgYjM2wnJypJNkor70gT1TAvB
         EAJQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:reply-to:date:in-reply-to:message-id
         :mime-version:references:subject:from:to:cc;
        bh=AxjHjFrYfKI0arB2F1ZWltZrEdR0g/4Qm9dbdrZzkxk=;
        b=iOvzzuxN3hpRGfJVnLr0yn2Ey4w74ZP/qhLPaCmhyHBp2DS9gCN8DknE3/bjTj+lwE
         tdY31jyTxysAZs7tEB+Vgc9KU5FP3zKipl4jDSaiISvgE4uMPk9FpgPHtIVF+WV6tmJy
         BTp9xLxycbv0Q4em5nqQZg7bZHfOWwDYh4DocphiFZfKdvtujWaNKKOqG+c8r7iVviqt
         FrXLGKCkhYwTS1MXSr9H3AzXZOaKS/7DGqRCs7yCjDEbCvlIvC/RLUf3nG98D9PbbIBw
         Vfb7wLIFOOODY3sGLUfQdw/8qQZ4Mi1o6G0oGkCkSkRT2dPwzXD2zp+Ly1osKZ/igCfE
         xjXQ==
X-Gm-Message-State: AOAM5326FgG3d2H2C4agSsSi9w5wgjTRYT8ULtK17JQLuTR/og8eYAJg
        pRlbPMJv/xyogDpcIRj99BFHvRgRsyw=
X-Google-Smtp-Source: 
 ABdhPJwMmyDI3xSTPxYqu+BbEtNr9UbWLJdYEeyFVIvBJrYrEWnpDSyGGibBiSxulYkloZvi2w2/pV6SgTM=
X-Received: from seanjc798194.pdx.corp.google.com
 ([2620:15c:f:10:6c6b:5d63:9b3b:4a77])
 (user=seanjc job=sendgmr) by 2002:a25:850a:: with SMTP id
 w10mr1756273ybk.402.1617160786960;
 Tue, 30 Mar 2021 20:19:46 -0700 (PDT)
Reply-To: Sean Christopherson <seanjc@google.com>
Date: Tue, 30 Mar 2021 20:19:35 -0700
In-Reply-To: <20210331031936.2495277-1-seanjc@google.com>
Message-Id: <20210331031936.2495277-3-seanjc@google.com>
Mime-Version: 1.0
References: <20210331031936.2495277-1-seanjc@google.com>
X-Mailer: git-send-email 2.31.0.291.g576ba9dcdaf-goog
Subject: [PATCH 2/3] KVM: SVM: Do not set sev->es_active until KVM_SEV_ES_INIT
 completes
From: Sean Christopherson <seanjc@google.com>
To: Paolo Bonzini <pbonzini@redhat.com>
Cc: Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org,
        Brijesh Singh <brijesh.singh@amd.com>,
        Tom Lendacky <thomas.lendacky@amd.com>
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Set sev->es_active only after the guts of KVM_SEV_ES_INIT succeeds.  If
the command fails, e.g. because SEV is already active or there are no
available ASIDs, then es_active will be left set even though the VM is
not fully SEV-ES capable.

Refactor the code so that "es_active" is passed on the stack instead of
being prematurely shoved into sev_info, both to avoid having to unwind
sev_info and so that it's more obvious what actually consumes es_active
in sev_guest_init() and its helpers.

Fixes: ad73109ae7ec ("KVM: SVM: Provide support to launch and run an SEV-ES guest")
Cc: stable@vger.kernel.org
Cc: Brijesh Singh <brijesh.singh@amd.com>
Cc: Tom Lendacky <thomas.lendacky@amd.com>
Signed-off-by: Sean Christopherson <seanjc@google.com>
---
 arch/x86/kvm/svm/sev.c | 29 ++++++++++++-----------------
 1 file changed, 12 insertions(+), 17 deletions(-)

diff --git a/arch/x86/kvm/svm/sev.c b/arch/x86/kvm/svm/sev.c
index 6481d7165701..97d42a007b86 100644
--- a/arch/x86/kvm/svm/sev.c
+++ b/arch/x86/kvm/svm/sev.c
@@ -87,7 +87,7 @@ static bool __sev_recycle_asids(int min_asid, int max_asid)
 	return true;
 }
 
-static int sev_asid_new(struct kvm_sev_info *sev)
+static int sev_asid_new(bool es_active)
 {
 	int pos, min_asid, max_asid;
 	bool retry = true;
@@ -98,8 +98,8 @@ static int sev_asid_new(struct kvm_sev_info *sev)
 	 * SEV-enabled guests must use asid from min_sev_asid to max_sev_asid.
 	 * SEV-ES-enabled guest can use from 1 to min_sev_asid - 1.
 	 */
-	min_asid = sev->es_active ? 0 : min_sev_asid - 1;
-	max_asid = sev->es_active ? min_sev_asid - 1 : max_sev_asid;
+	min_asid = es_active ? 0 : min_sev_asid - 1;
+	max_asid = es_active ? min_sev_asid - 1 : max_sev_asid;
 again:
 	pos = find_next_zero_bit(sev_asid_bitmap, max_sev_asid, min_asid);
 	if (pos >= max_asid) {
@@ -179,13 +179,14 @@ static void sev_unbind_asid(struct kvm *kvm, unsigned int handle)
 static int sev_guest_init(struct kvm *kvm, struct kvm_sev_cmd *argp)
 {
 	struct kvm_sev_info *sev = &to_kvm_svm(kvm)->sev_info;
+	bool es_active = argp->id == KVM_SEV_ES_INIT;
 	int asid, ret;
 
 	ret = -EBUSY;
 	if (unlikely(sev->active))
 		return ret;
 
-	asid = sev_asid_new(sev);
+	asid = sev_asid_new(es_active);
 	if (asid < 0)
 		return ret;
 
@@ -194,6 +195,7 @@ static int sev_guest_init(struct kvm *kvm, struct kvm_sev_cmd *argp)
 		goto e_free;
 
 	sev->active = true;
+	sev->es_active = es_active;
 	sev->asid = asid;
 	INIT_LIST_HEAD(&sev->regions_list);
 
@@ -204,16 +206,6 @@ static int sev_guest_init(struct kvm *kvm, struct kvm_sev_cmd *argp)
 	return ret;
 }
 
-static int sev_es_guest_init(struct kvm *kvm, struct kvm_sev_cmd *argp)
-{
-	if (!sev_es)
-		return -ENOTTY;
-
-	to_kvm_svm(kvm)->sev_info.es_active = true;
-
-	return sev_guest_init(kvm, argp);
-}
-
 static int sev_bind_asid(struct kvm *kvm, unsigned int handle, int *error)
 {
 	struct sev_data_activate *data;
@@ -1128,12 +1120,15 @@ int svm_mem_enc_op(struct kvm *kvm, void __user *argp)
 	mutex_lock(&kvm->lock);
 
 	switch (sev_cmd.id) {
+	case KVM_SEV_ES_INIT:
+		if (!sev_es) {
+			r = -ENOTTY;
+			goto out;
+		}
+		fallthrough;
 	case KVM_SEV_INIT:
 		r = sev_guest_init(kvm, &sev_cmd);
 		break;
-	case KVM_SEV_ES_INIT:
-		r = sev_es_guest_init(kvm, &sev_cmd);
-		break;
 	case KVM_SEV_LAUNCH_START:
 		r = sev_launch_start(kvm, &sev_cmd);
 		break;

From patchwork Wed Mar 31 03:19:36 2021
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Sean Christopherson <seanjc@google.com>
X-Patchwork-Id: 12174265
Return-Path: <kvm-owner@kernel.org>
X-Spam-Checker-Version: SpamAssassin 3.4.0 (2014-02-07) on
	aws-us-west-2-korg-lkml-1.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-26.3 required=3.0 tests=BAYES_00,DKIMWL_WL_MED,
	DKIM_SIGNED,DKIM_VALID,DKIM_VALID_AU,HEADER_FROM_DIFFERENT_DOMAINS,
	INCLUDES_CR_TRAILER,INCLUDES_PATCH,MAILING_LIST_MULTI,SPF_HELO_NONE,SPF_PASS,
	USER_AGENT_GIT,USER_IN_DEF_DKIM_WL autolearn=unavailable autolearn_force=no
	version=3.4.0
Received: from mail.kernel.org (mail.kernel.org [198.145.29.99])
	by smtp.lore.kernel.org (Postfix) with ESMTP id 7836BC433E1
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 03:20:53 +0000 (UTC)
Received: from vger.kernel.org (vger.kernel.org [23.128.96.18])
	by mail.kernel.org (Postfix) with ESMTP id 51524619E5
	for <kvm@archiver.kernel.org>; Wed, 31 Mar 2021 03:20:53 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S233609AbhCaDUV (ORCPT <rfc822;kvm@archiver.kernel.org>);
        Tue, 30 Mar 2021 23:20:21 -0400
Received: from lindbergh.monkeyblade.net ([23.128.96.19]:42368 "EHLO
        lindbergh.monkeyblade.net" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S233492AbhCaDTu (ORCPT <rfc822;kvm@vger.kernel.org>);
        Tue, 30 Mar 2021 23:19:50 -0400
Received: from mail-yb1-xb4a.google.com (mail-yb1-xb4a.google.com
 [IPv6:2607:f8b0:4864:20::b4a])
        by lindbergh.monkeyblade.net (Postfix) with ESMTPS id 1820AC06175F
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 20:19:50 -0700 (PDT)
Received: by mail-yb1-xb4a.google.com with SMTP id k189so773577ybb.17
        for <kvm@vger.kernel.org>; Tue, 30 Mar 2021 20:19:50 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=reply-to:date:in-reply-to:message-id:mime-version:references
         :subject:from:to:cc;
        bh=3Xa1OKEIcZ5edRwAMZLibB+Ad1pcD0ADTTrJiSO76FU=;
        b=U0dDAMw5AWHxlOqCtiXJpoymuGVxzLFXL1A4mlfI9uLaMbLLGk45tLAgKX1qjOccmR
         cTevpLRzVJqNmfwPQFzh9m2DRDgvyV3sxKLBL7DGw0tBxNZU1FmmCrBGRaxm+XeF9GhC
         Hwe2L+EpDVfyH3o2QCwSMd12qVhUUklccyj9paEDibFrU+0bp42pkhunV4PyqQJqGCCY
         lDKuGej38L4ReQUgV+4kMcPpL8C5+/EMh7xgg98CsQBhDcKL9WWhhvb8H9cbfufK/VfC
         IkdR/BrdtTxnTKbkHjjcbWTKDtDWNRmft5OnlBUgBnLeNFPVsA6t6+1wUWKw4E/Jv7Wj
         rXUQ==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:reply-to:date:in-reply-to:message-id
         :mime-version:references:subject:from:to:cc;
        bh=3Xa1OKEIcZ5edRwAMZLibB+Ad1pcD0ADTTrJiSO76FU=;
        b=b1UGm8QkQi0YZxyMQntZ2d8rVu+eQD9zB5DItho1m1MrxIwkTpFfipOujvGPA9gaKW
         eh09m2vBF954YolROVwFmZ1ltZ2kIgLN6smpGVwDm+PndHOk5QyvY8dX8JVkAYvCNg9r
         YQmQ4lUr1nQbJ3SSrGBZZzh8VqaLMfdBuCwkEnZQ7qPIrbAyoT58uiUIV1I4e/4RRk6V
         pD+19R2yMFBJ+cQcq6UFZKSB3tOcRMqJxMHYhojdt2sZ489pqIdURVT0phYgvRvThOOj
         0gx4ngeBHDNXVwdg5uAQMwoy0betvRGqR3ovEpukg6dHqUg+sHNDTLU8c8v6eXOzFg0x
         XGJQ==
X-Gm-Message-State: AOAM5310oAULsUuubs+iz1EYsDXYWIIj9R6NvBHzvJC7FK8eXuTtvweb
        u5QirzbXbCPlz6ZanIvowT5Xum9Iy3I=
X-Google-Smtp-Source: 
 ABdhPJzBx7uD+67poCDf2QYRxAznyzYMGF0Hynxyll63eq1q/oZtnzx23zwvh/FRi1trMJENU/uX+ZtWzOA=
X-Received: from seanjc798194.pdx.corp.google.com
 ([2620:15c:f:10:6c6b:5d63:9b3b:4a77])
 (user=seanjc job=sendgmr) by 2002:a25:268e:: with SMTP id
 m136mr1876235ybm.220.1617160789338;
 Tue, 30 Mar 2021 20:19:49 -0700 (PDT)
Reply-To: Sean Christopherson <seanjc@google.com>
Date: Tue, 30 Mar 2021 20:19:36 -0700
In-Reply-To: <20210331031936.2495277-1-seanjc@google.com>
Message-Id: <20210331031936.2495277-4-seanjc@google.com>
Mime-Version: 1.0
References: <20210331031936.2495277-1-seanjc@google.com>
X-Mailer: git-send-email 2.31.0.291.g576ba9dcdaf-goog
Subject: [PATCH 3/3] KVM: SVM: Do not allow SEV/SEV-ES initialization after
 vCPUs are created
From: Sean Christopherson <seanjc@google.com>
To: Paolo Bonzini <pbonzini@redhat.com>
Cc: Sean Christopherson <seanjc@google.com>,
        Vitaly Kuznetsov <vkuznets@redhat.com>,
        Wanpeng Li <wanpengli@tencent.com>,
        Jim Mattson <jmattson@google.com>,
        Joerg Roedel <joro@8bytes.org>, kvm@vger.kernel.org,
        linux-kernel@vger.kernel.org,
        Brijesh Singh <brijesh.singh@amd.com>,
        Tom Lendacky <thomas.lendacky@amd.com>
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org

Reject KVM_SEV_INIT and KVM_SEV_ES_INIT if they are attempted after one
or more vCPUs have been created.  KVM assumes a VM is tagged SEV/SEV-ES
prior to vCPU creation, e.g. init_vmcb() needs to mark the VMCB as SEV
enabled, and svm_create_vcpu() needs to allocate the VMSA.  At best,
creating vCPUs before SEV/SEV-ES init will lead to unexpected errors
and/or behavior, and at worst it will crash the host, e.g.
sev_launch_update_vmsa() will dereference a null svm->vmsa pointer.

Fixes: 1654efcbc431 ("KVM: SVM: Add KVM_SEV_INIT command")
Fixes: ad73109ae7ec ("KVM: SVM: Provide support to launch and run an SEV-ES guest")
Cc: stable@vger.kernel.org
Cc: Brijesh Singh <brijesh.singh@amd.com>
Cc: Tom Lendacky <thomas.lendacky@amd.com>
Signed-off-by: Sean Christopherson <seanjc@google.com>
---
 arch/x86/kvm/svm/sev.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/arch/x86/kvm/svm/sev.c b/arch/x86/kvm/svm/sev.c
index 97d42a007b86..824bc7d22e77 100644
--- a/arch/x86/kvm/svm/sev.c
+++ b/arch/x86/kvm/svm/sev.c
@@ -182,6 +182,9 @@ static int sev_guest_init(struct kvm *kvm, struct kvm_sev_cmd *argp)
 	bool es_active = argp->id == KVM_SEV_ES_INIT;
 	int asid, ret;
 
+	if (kvm->created_vcpus)
+		return -EINVAL;
+
 	ret = -EBUSY;
 	if (unlikely(sev->active))
 		return ret;
